### ISSUE TITLE
object_detection A problem (bug?) when training a detection model in the multiclass setting

### ISSUE THREAD

**Issue by kolesman** [tensorflow/models#3132]
### System information
- **What is the top-level directory of the model you are using**: research/object_detection
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Debian 4.9.65
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: ('v1.4.0-19-ga52c8d9', '1.4.1')
- **Bazel version (if compiling from source)**: 
- **CUDA/cuDNN version**:  8.0 / 6
- **GPU model and memory**: P100 16GB
- **Exact command to reproduce**: N/A

### Describe the problem
I use the object detection API to fine-tune the faster-rcnn detection model on
my custom dataset. The dataset often has multiple identical object boxes with
different labels. Thus, I use sigmoid score converter, sigmoid classification
loss and enable "merge_multiple_label_boxes: true" in the config file. Unfortunately,
I get very poor model in the end. The box labels are generally correct, but their locations
and confidences are poor.

I suspect that there is either something wrong with my configuration file or there is
a bug in how multiclass detection datasets are handled.

I can reliably reproduce the wrong behavior by enabling the option "merge_multiple_label_boxes: true"
for multi-class datasets. In other words, the problem appears when a *k-hot label encoding* is used.

I verified that, when using the sigmoid score converter and the sigmoid
loss, a detection model works fine with 1-hot label encoding (i.e. unique labels per box).
Thus, I presume that k-hot encoding breaks the detection pipeline, but I am
not able to find the exact reason.

In the following I describe a minimal example for reproducing the error.

### Source code / logs
I've created a fork with minimal changes, which allows to reproduce the problem: https://github.com/kolesman/models

Specifically, I implement the following changes:

1. Modify the *create_pet_tf_record.py* to produce multiclass dataset.
    A box labeled by a class $c$ is populated by labels $c + 1$ and $c + 2$.

2. Modify the sample config file faster_rcnn_resnet101_pets.config by
      -> enabling sigmoid score converter
      -> enabling sigmoid loss
      -> enabling "merge_multiple_label_boxes: true"

Otherwise, I closely follow the repository instructions, use mscoco pretrained model and
employ train.py and eval.py scripts:
```bash

# from research/
python object_detection/dataset_tools/create_pet_tf_record.py --label_map_path=object_detection/data/pet_label_map.pbtxt --data_dir=`pwd` --output_dir=`pwd`

# from research/object_detection
python train.py --logtostderr --pipeline_config_path=samples/configs/faster_rcnn_resnet101_pets.config --train_dir=train_dir

# from research/object_detection
python eval.py --logtostderr --pipeline_config_path=samples/configs/faster_rcnn_resnet101_pets.config --checkpoint_dir=train_dir/ --eval_dir=eval_dir

```

The resulting detection outputs after more then 10.000 iterations of training are attached here.
![multiclass](https://user-images.githubusercontent.com/460828/34731442-a28e50e2-f562-11e7-8838-15fb963ac735.png)

The labels are generally correct, but the boxes and their confidences are very poor.
Interestingly, quality of boxes degrades over the course of training.

**Comment by cy89** [comment]
@afathi3 @dreamdragon @jesu9 are you aware of any problems with k-hot label encoding?

**Comment by kolesman** [comment]
I think there is actually a bug and it can be easily fixed.

The potential problem is in this line: https://github.com/tensorflow/models/blob/9b57f41ce21cd7264c52140c9ab31cdfc5169fcd/research/object_detection/meta_architectures/faster_rcnn_meta_arch.py#L1476

Essentially, it always prefers classes with the lowest number for constructing the second stage regression loss. In the multiclass setting it results in a situation, when the classes with higher numbers are neglected.  

The solution is to add small noise to the label matrix in order to pick a class at random: https://github.com/tensorflow/models/compare/master...kolesman:multiclass-fix?expand=1
This simple modification seems to fix the major problem for me.

 Big thanks to Jasper Uijlings for the helpful discussion, which helped me to locate the bug.

**Comment by afathi3** [comment]
Adding @jch1 and @tombstone

**Comment by jrruijli** [comment]
I agree with Alex that his solution would work. However, it would probably be better (and much harder) to assign the regression to all classes but properly weight them. See also rationale for introducing this bug in the first place:

# For anchors with multiple labels, picks refined_location_encodings
# for just one class to avoid over-counting for regression loss and
# (optionally) mask loss.

Cheers,
Jasper

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
